### PR TITLE
Java: Promote JNDI Injection query from experimental

### PR DESCRIPTION
**Pull request by atorralba**
PR to promote the JNDI Injection query created in https://github.com/github/codeql/pull/3288

## Changes

* Existing files were moved out of experimental.
* The `JndiInjectionLib.qll` file was renamed to `JndiInjection.qll` and refactored to use the CSV sink model. Also, added new sinks.
* No longer necessary `.qll` libs were removed from experimental. `frameworks/Jndi.qll` was kept, since it contains classes still in use.
* Refactored tests to use `InlineExpectationsTest`.

## To consider

* `UnsafeSearchControlsSink` is probably worth reviewing - it uses a `DataFlow::Configuration` to detect new sinks receiving a `SearchControls` parameter with `.setReturningObjFlag(true)`

### CODE REVIEW

**Comment by atorralba** [comment]
@github/docs-content-codeql please review the qhelp file. Even though changes aren't introduced in this PR, it wasn't reviewed when this query was merged to experimental.

File: java/ql/src/Security/CWE/CWE-074/JndiInjection.ql [hunk 1]
```diff
@@ -11,8 +11,7 @@
  */
 
 import java
-import semmle.code.java.dataflow.FlowSources
-import JndiInjectionLib
+import semmle.code.java.security.JndiInjectionQuery
```
Comment (mchammer01) [inline java/ql/src/Security/CWE/CWE-074/JndiInjection.ql:14]: This comment is for line 3 (query description)

Suggesting the following:

Performing a JNDI lookup with a user-controlled name can lead to the download of an untrusted object and to execution of arbitrary code.

**Review by mchammer01 — COMMENTED** [review]
@atorralba - this is looking great 💖 
I made a minor update in the qhelp file, and made a suggestion to improve the query description.
Hope this helps!

File: java/ql/src/Security/CWE/CWE-074/JndiInjection.ql (hunk 1 above)
Comment (atorralba) [reply java/ql/src/Security/CWE/CWE-074/JndiInjection.ql:14]: Thank you for the review @mchammer01! I applied your suggestion in 29490e5.

File: java/ql/test/query-tests/security/CWE-074/JndiInjectionTest.ql [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.FlowSources
```
Comment (aschackmull) [inline java/ql/test/query-tests/security/CWE-074/JndiInjectionTest.ql:3]: Superfluous, right?
```suggestion
```

File: java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+private class UnsafeSearchControlsConf extends DataFlow2::Configuration {
+  UnsafeSearchControlsConf() { this = "UnsafeSearchControlsConf" }
+
+  override predicate isSource(DataFlow2::Node source) { source instanceof UnsafeSearchControls }
+
+  override predicate isSink(DataFlow2::Node sink) { sink instanceof UnsafeSearchControlsArgument }
+}
+
+/**
+ * An argument of type `SearchControls` of a a `LdapOperations.search` or `DirContext.search` call.
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll:54]:
```suggestion
 * An argument of type `SearchControls` of an `LdapOperations.search` or `DirContext.search` call.
```

File: java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      conf.hasFlowTo(DataFlow::exprNode(ma.getAnArgument()))
+    |
+      this.asExpr() = ma.getArgument(0)
+    )
+  }
+}
+
+/**
+ * Find flows between a `SearchControls` object with `setReturningObjFlag` = `true`
+ * and an argument of a `LdapOperations.search` or `DirContext.search` call.
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll:43]:
```suggestion
 * and an argument of an `LdapOperations.search` or `DirContext.search` call.
```

File: java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ */
+private class UnsafeSearchControlsConf extends DataFlow2::Configuration {
+  UnsafeSearchControlsConf() { this = "UnsafeSearchControlsConf" }
+
+  override predicate isSource(DataFlow2::Node source) { source instanceof UnsafeSearchControls }
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll:48]: The `Node` type is shared between different dataflow instances, so might as well use the simpler name:
```suggestion
  override predicate isSource(DataFlow::Node source) { source instanceof UnsafeSearchControls }
```

File: java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+  override predicate isSink(DataFlow2::Node sink) { sink instanceof UnsafeSearchControlsArgument }
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll:50]:
```suggestion
  override predicate isSink(DataFlow::Node sink) { sink instanceof UnsafeSearchControlsArgument }
```

File: java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+/**
+ * An argument of type `SearchControls` of a a `LdapOperations.search` or `DirContext.search` call.
+ */
+private class UnsafeSearchControlsArgument extends DataFlow2::ExprNode {
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll:56]:
```suggestion
private class UnsafeSearchControlsArgument extends DataFlow::ExprNode {
```

File: java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  SetReturningObjFlagMethod() {
+    this.getDeclaringType() instanceof TypeSearchControls and
+    this.hasName("setReturningObjFlag")
+  }
+}
+
+/** The class `java.util.Hashtable`. */
+private class TypeHashtable extends Class {
+  TypeHashtable() { this.getSourceDeclaration().hasQualifiedName("java.util", "Hashtable") }
+}
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll:102]: Unused?
```suggestion
```

**Review by aschackmull — COMMENTED** [review]
Some minor comments, otherwise LGTM.

File: java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll (hunk 8 above)
Comment (atorralba) [reply java/ql/src/semmle/code/java/security/JndiInjectionQuery.qll:102]: Yes, it was copied to `JndiInjection.qll` but the intention was to move it instead. Thanks!

File: java/ql/test/query-tests/security/CWE-074/JndiInjectionTest.ql (hunk 2 above)
Comment (atorralba) [reply java/ql/test/query-tests/security/CWE-074/JndiInjectionTest.ql:3]: Indeed, those should have been removed when `JndiInjectionFlowConfig` was reused here. Good catch.

**Review by aschackmull — APPROVED** [review]

**Pull request merged**